### ISSUE TITLE
RFC 0019: Updating to JWE compliant data model

### ISSUE THREAD

**Issue by kdenhartog** [hyperledger/aries-rfcs#133]
Recently there's been discussion to update this work to be fully JWE compliant. One of the proposed solutions that's come up is to generate the following format:

```json
{
    "protected": base64url({
        "typ": "prs.hyperledger.aries-auth-message",
        "alg": "ECDH+XC20PKW",
        "enc":"XC20P"
    }),
    "recipients": [
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "kid": "base58(Recipient vk)",
                "iv": "base64url(CEK encryption IV)",
                "pk": "compactJWE(Sender PK)"
            }
        },
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "kid": "base58(Recipient vk)",
                "iv": "base64url(CEK encryption IV)",
                "pk": "compactJWE(Sender PK)"
            }
        }
    ],
    "aad": "base64url(sha256(concat('.',sort([recipients[0].kid, recipients[n].kid]))))",
    "iv": "base64url(content encryption IV)",
    "ciphertext": "base64url(ciphertext)",
    "tag": "base64url(AEAD Authentication Tag)"
}

Exploded example of compactJWE(Sender PK):
{
    "protected": base64url({
        "iv": "base64url(CEK encryption IV)",
        "epk": "Ephemeral JWK",
        "typ": "jose",
        "cty": "jwk+json",
        "alg": "ECDH-ES+XC20PKW",
        "enc":"XC20P"
    }),
    "encrypted_key": "base64url(encrypted CEK)",
    "iv": "base64url(content encryption IV)",
    "ciphertext": "base64url(Encrypted Sender vk as JWK)",
    "tag": "base64url(AEAD Authentication Tag)"
}
```

Here's a provided example of what this would look like:
```json
{
    "protected": "eyJ0eXAiOiJwcnMuaHlwZXJsZWRnZXIuYXJpZXMtYXV0aC1tZXNzYWdlIiwiYWxnIjoiRUNESCtYQzIwUEtXIiwiZW5jIjoiWEMyMFAifQ",
    "recipients": [
        {
            "encrypted_key": "whpkJkvHRP0XX-EqxUOHhHIfuW8i5EMuR3Kxlg5NNIU",
            "header": {
                "kid": "5jMonJACEPcLfqVaz8jpqBLXHHKYgCE71XYBmFXhjZVX",
                "iv": "tjGLK6uChZatAyACFzGmFR4V9othKN8S",
                "tag": "ma9pIjkQuzaqvq_5y5vUlQ",
                "pk": "eyJpdiI6IldoVGptNi1DX2xiTlQ4Q2RzN2dfNjdMZzZKSEF3NU5BIiwiZXBrIjp7Imt0eSI6Ik9LUCIsImNydiI6IlgyNTUxOSIsIngiOiJnNjRfblJSSFQyYk1JX1hjT1dHRTdJOGdQcU1VWTF4aUNub2J0LVhDUkNZIn0sInR5cCI6Impvc2UiLCJjdHkiOiJqd2sranNvbiIsImFsZyI6IkVDREgtRVMrWEMyMFBLVyIsImVuYyI6IlhDMjBQIn0.4zUt5tOOlcQWskJqxfMi0tNsfUCAzb5_PDfPqQ1h0Vw.xYkeEXV1_cSYFEd6UBMIfl8MWQfHaDex.XSNKTRXye5-iSXQ-aS_vQVZNEgFE6iA9X_KgSRMzihQBMoI1j4WM3o-9dMT9TeSyMvdq3gXt1NpvLdZHpJplahhk3mxMZL-vawm5Prtf.H7a5N-dggwdesjHyJCl06w"
            }
        },
        {
            "encrypted_key": "dDHydlp_wlGt_zwR-yUvESx9fXuO-GRJFGtaw2u6CEw",
            "header": {
                "kid": "TfVVqzPT1FQHdq1CUDe9XYcg6Wu2QMusWKhGBXEZsosg",
                "iv": "7SFlGTxQ4Q2l02D9HRNdFeYQnwntyctb",
                "tag": "9-O6djpNAizix-ZnjAx-Fg",
                "pk": "eyJpdiI6IkV6UjBFaVRLazJCT19oc05qOVRxeU9PVmVLRFFPYVp1IiwiZXBrIjp7Imt0eSI6Ik9LUCIsImNydiI6IlgyNTUxOSIsIngiOiJoU1g1NGt5ZTdsd0pBdjlMaUplTmh4eFhaV1N0M3hLSDBXUmh6T1NOb1c0In0sInR5cCI6Impvc2UiLCJjdHkiOiJqd2sranNvbiIsImFsZyI6IkVDREgtRVMrWEMyMFBLVyIsImVuYyI6IlhDMjBQIn0.qKmU5xO8Z1ZtRBWEjEMixb5VZG0mrY0LnjUGjLqktwg.EG-VOZSC2vLdoO5l2_A37IYvdXCckLZp.D4kgD6bYL1YfXyApk5ESKE2sc8TUiO-QGBtY-M5hcV_F88JPZdsi53Qofxk02ZxPHJZK-abDy45pIMH6-KUMDfE6WKhW3nPQhydPYutv.0SO4VjM8sDH-wGHcEpinTg"
            }
        }
    ],
    "aad": "OGY5ZDIxMDE3YTQ4MTc4YWE5MTk0MWQyOGJmYjQ1ZmZmMTYzYTE3ZjUxYjc4YjA3YTlmY2FlMmMwOTFlMjBhZg",
    "ciphertext": "x1lnQq_pZLgU2ZC4",
    "tag": "2JgOe9SRjJXddT9TyIjqrg",
    "iv": "fDGEXswlWXOBx6FxPC_u6qIuhADnOrW1"
}
```

I'm curious what other's opinions are on this approach. I personally am not a fan of the bloat the compact JWE adds in order to encrypt the sender's public key. However, I've not found another approach yet that seems satisfactory. If anyone has some suggestions it would be appreciated.

**Comment by kdenhartog** [comment]
Please be aware, this issue is also posted on the [DIDComm-js library](https://github.com/decentralized-identity/DIDComm-js/issues/8) in DIF. I'll facilitate communication across the groups if you don't want to track both conversations.

**Comment by troyronda** [comment]
I would like to see Aries using a JWE compliant data model.

**Comment by Baha-sk** [comment]
the above example is for Authrcrypt, below is the Anoncrypt example: 
`{
    "protected": base64url({
        "typ": "prs.hyperledger.aries-anon-message",
        "alg": "ECDH-ES+XC20PKW",
        "enc":"XC20P"
    }),
    "recipients": [
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "kid": "base64url(Recipient vk)",
                "iv": "base64url(CEK encryption IV)",
                "epk": "Ephemeral JWK",
            }
        },
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "kid": "base64url(Recipient vk)",
                "iv": "base64url(CEK encryption IV)",
                "epk": "Ephemeral JWK",
            }
        }
    ],
    "aad": "base64url(sha256(concat('.',sort([recipients[0].kid, recipients[n].kid]))))",
    "iv": "base64url(content encryption IV)",
    "ciphertext": "base64url(ciphertext)",
    "tag": "base64url(AEAD Authentication Tag)"
}`

**Comment by llorllale** [comment]
It's worth pointing to the internet draft that originally defined the JOSE AEAD algorithms: https://tools.ietf.org/html/draft-amringer-jose-chacha-00

**Comment by gamringer** [comment]
I have just updated the draft to include ECDH-ES bits:

https://datatracker.ietf.org/doc/draft-amringer-jose-chacha/

**Comment by kdenhartog** [comment]
@gamringer I'm not finding documentation or an IANA registry of the `pk` field. Do you have details about it?

**Comment by gamringer** [comment]
I'm writing up the I-D for ECDH for JOSE separately from the Chacha for JOSE I-D. I'm writing it up to have the `pk` field be the Sender's JWK, or JWE-envelopped JWK, as was the case in my example.

**Comment by kdenhartog** [comment]
Ok, since we're registering the `pk` field is it possible that we can try and find a more compact form for this? With ed25519 and secp256k1 keys I'm not too concerned about size (it's not great but at least manageable), but with a 3072 bit RSA key, this is going to going bloat the message to the point where it takes a second or 2 to load the messages and could make it not possible to embed in QR codes.

**Comment by gamringer** [comment]
Then we can register both `opk`, and `oid` to serve are mirror to recipient versions where `opk` is to `jwk` what `oid` is to `kid`. That way, `prs.hyperledger.aries-auth-message` can just use `oid` to hold a `crypto_box_seal` of the sender's key identifier, since for format for `oid` (as for `kid`) is application dependant.

**Comment by kdenhartog** [comment]
That seems like a good solution that will allow us to work with multiple types then.

**Comment by Baha-sk** [comment]
@gamringer @kdenhartog I pushed an animal PR to our go-jose fork to support the above format using Go:
unit tests are passing: https://github.com/trustbloc/go-jose/pull/1
your input is appreciated.

**Comment by Baha-sk** [comment]
> That seems like a good solution that will allow us to work with multiple types then.

@gamringer @kdenhartog can we have an updated JWE format with `opk` and `oid` examples? I wanna be clear on what fields our JWE will be using to reference these key IDs.

**Comment by gamringer** [comment]
I finished reimplementing my proposal in PHP: https://github.com/gamringer/php-authcrypt

**Comment by Baha-sk** [comment]
I believe we should update the format to meet the standards and reference them in the docs.

@gamringer, I gathered these links from the discussion we had such as:

- Key derivation for ECDH:

https://tools.ietf.org/html/rfc7518#section-4.6.2

More specifically about the alternative format for `apu` with a random 64 byte value and no `apk`:
```Alternatively, applications MAY conduct key derivation in
   a manner similar to "Diffie-Hellman Key Agreement Method" [RFC2631]:
   in that case, the "apu" parameter MAY either be omitted or represent
   a random 512-bit value (analogous to PartyAInfo in Ephemeral-Static
   mode in RFC 2631) and the "apv" parameter SHOULD NOT be present.

   See Appendix C for an example key agreement computation using this
   method.```

- mapping of new key identifier fields (replacing `pk`): https://tools.ietf.org/html/rfc8152#section-12.4.1. @gamringer I'm quoting your specific change:
```instead of `opk` and `oid`, I will change it to `spk` and `sid```

we should probably update the format to reflect these changes

**Comment by Baha-sk** [comment]
@kdenhartog @gamringer, I just pushed the Go implementation of the JWE encryption (decryption will be coming up soon). Here is an example format:
Encryption with XC20P: 
```
{
            	"protected": {
            		"alg": "ECDH-SS+XC20PKW",
            		"enc": "XC20P",
            		"typ": "prs.hyperledger.aries-auth-message"
            	},
            	"recipients": [
            		{
            			"encrypted_key": "C1ZgZxgIKne86FlGuct0L2y2UENtHyOS78KvoVgBAaM=",
            			"header": {
            				"apu": "dpQnrrNMi8CHC1sRLxcKdzHUfYqM9hMI4P8VpLMArVYdJQlmnIzF6niGjzTCNF6UBobaTmm6msTlEK2chntsOA==",
            				"iv": "Po7Wi7CKWHeCof450oyrFB9A_x8D8LQa",
            				"tag": "QkkS6QqQLaWRJDtEK0htKQ==",
            				"kid": "ETu2KkCZUawEHq99z2RubyCsnVyaLhzMiYTHMu3uVncp",
            				"oid": "PBqOfOOkMM3hDUxSunySlld6-15qdbMdWajolmmKVYs0bog785bnqckvoPYDkqoMXwVPtr-qPQgIpg9A"
            			}
            		},
            		{
            			"encrypted_key": "pGL8MBT1WUXjfXuwSl2enj5N1BLtmeEZ1wAZAzUoH60=",
            			"header": {
            				"apu": "iSa2FM-_nOzxG4IjX5MOUv1uvEFs-nA-fYlrLry66VVrNJxakXclPC7UnJ_RtqI_cwlEgJ_GaVHcNv84NA8VjQ==",
            				"iv": "xSEHpsn1iuy4x3KTfeHdcdBRziqCvImG",
            				"tag": "vsU_H6K5m4_jVFWNxxV--A==",
            				"kid": "BkcdRt9AChfFXGvKU3ecjzyNCzoP4EBVvm5dtgvqv5UX",
            				"oid": "u1PbWv2wIIV4BkOibItT4DzkzVb6d75hLhS01Ha1CzapUX5xjBlcyjwrl6nIOrARBvoJyH0FXZdkxec7"
            			}
            		},
            		{
            			"encrypted_key": "PsFDcK7-VHzH-E-KDC-Cy1NHZYKepv1X7zolbuRbdbw=",
            			"header": {
            				"apu": "PAYUGCNWRKGq82HvOxyMVE6VI13mYlvowK5X3NqZsFiwqpTTwXgNTZA5Tnxch8vsBZAzeEpmEvI3M9YpJflnrA==",
            				"iv": "s_FfFvulVhFKwCoalqp9qESsEtWnDgZ9",
            				"tag": "7IIQuM9Hs2Zqsj9Vga2nAQ==",
            				"kid": "EikHy4Hmbw5zGUMPGYDoAwYFNKnrKeDQkFqLxMvTXD83",
            				"oid": "EaM4cO2c6AIMWBDwOnQSSfc95qHoF1_8OesVSrK2-O1RLWTEq3QB0cnj46WIa_KIDNq-wpwxRmNQu3-5"
            			}
            		}
            	],
            	"aad": "itXVSOmyKyNvjHYSDiCpvxB0HqlFfy5YJtdQftWqjEk=",
            	"iv": "WsiI5jXbXKYUJjwF0H3ZVl02meBL26bL",
            	"tag": "jMmIt9MsYRtboMwA4yrf2g==",
            	"ciphertext": "VbnjyXkgej1Ix1gjxkVs0ncGWauOxyt5QzY="
            }
```

Go PR: https://github.com/hyperledger/aries-framework-go/pull/107


as mentioned in my previous post, we need to update the format a little and confirm if we're using `oid` or `sid` (or any other fields)

**Comment by kdenhartog** [comment]
Update on this. I've been looking at this recently and trying to weigh all factors that would get us good interop between all parties in DID community. Since this change would create a breaking change, I want to make sure that what ever we move to accounts for all of our needs for quite some time.

For example, one thing that I didn't account for that I'd like to in the future is the ability to add in forward secrecy when we perform this breaking change.

Since I know this work is blocking @Baha-sk my recommendation is to implement the currently non-compliant JWE for now and we'll say that it is version 1. This will give us time to figure out what we want and to conduct a larger discussion as to what we should do going forward. Does that feel like a reasonable approach?

**Comment by troyronda** [comment]
We would like to use a compliant JWE. We have already started implementing this updated approach.

No - I cannot say dropping the compliant version is reasonable.

**Comment by Baha-sk** [comment]
Thanks for the update @kdenhartog 

I do agree with @troyronda 

There has been some effort done to make it JWE compliant, it makes sense to finish off the work and be JWA compliant. The sooner we deliver this the better (before v1) to avoid backward compatibility issues between agents.

**Comment by kdenhartog** [comment]
After our discussion on the call today about key identifiers, I was convinced that the explicitness of JWKs will be more beneficial then the conciseness of the libsodium option. @gamringer do you still have to data model that you originally described like this? Would we need to change anything to make that the standard implementation?

**Comment by gamringer** [comment]
@kdenhartog I did not reimplement that version of my proposal in a poc after my workstation failure. I can take a couple of hours to do it.

**Comment by Baha-sk** [comment]
> After our discussion on the call today about key identifiers, I was convinced that the explicitness of JWKs will be more beneficial then the conciseness of the libsodium option. @gamringer do you still have to data model that you originally described like this? Would we need to change anything to make that the standard implementation?

@kdenhartog is this confirmed to be the final change about the JWE encryption? Because I believe it's simpler to just box.Open the base58 decoded `OID` with the recipient's private key than having to JWE unmarshal `spk` then extract the key from the jwk (it's less work).

Mind you, none of this is part of the JWE standard yet, so we still have a chance at making it simpler.

what do you think?

**Comment by llorllale** [comment]
@kdenhartog can you elaborate on why we should not include a standard `sender` key in DIDComm plaintext messages?

**Comment by troyronda** [comment]
@Baha-sk I believe we should make our design choices based on what is most consistent with the existing specs (regardless of extra work). I would assume wrapping a key in a JWK would be more consistent...

**Comment by Baha-sk** [comment]
I guess we want to stick to JWE/JWK format as much as possible @kdenhartog , this is why we want the sender's key to be in JWK format?

**Comment by llorllale** [comment]
@kdenhartog reminder: the discussion on the last call revolved around two points: 1) how to encode the inline key such that its format can be interpreted, and 2) how to identify the Sender of the DIDComm msg.

It is understood that JWKs solve problem 1. Are you proposing they also solve problem 2?

**Comment by kdenhartog** [comment]
> @kdenhartog reminder: the discussion on the last call revolved around two points: 1) how to encode the inline key such that its format can be interpreted, and 2) how to identify the Sender of the DIDComm msg.
> 
> It is understood that JWKs solve problem 1. Are you proposing to also solve problem 2 with them?

Sorry, I've been away at RWoT, but was able to do some investigation into this. I do believe that JWKs would allow us to solve both problems. The JWK format provides a `kid` field which can list the DID URI, while also containing all the necessary cryptographic material in the other parts of the JWK format.

@selfissued Would you mind taking a look at this discussion and validating the direction we're going as well? This is the discussion I mentioned to you at RWoT.

**Comment by dhh1128** [comment]
I hadn't been very aware of this conversation (although I knew it was happening in theory). I wanted to articulate two requirements that I consider important:

1. **Preserve the option to operate in repudiable mode.** (This is an absolute dealbreaker for me; if we lose this, I can't use the solution, period. I believe from scanning the thread and from personal conversations with Kyle that this requirement is well understood and we are tracking it to be satisfied by the final solution, so me articulating it may be unnecessary--but I'm just checking.)

2. **Solve the problem of runaway bloat caused by repeated re-encryption.** (This is NOT an absolute dealbreaker for me, but is highly desirable. We proposed a few months back to change the encryption envelope to address this, but the proposal was deferred because we didn't think it was worth it to modify the envelope format solely to add this feature. Now that we're considering a change to the format for different reasons, I strongly prefer that we tackle this problem at the same time, and not defer it again.)

Am I correct that requirement #1 is taken care of? And where are we at with requirement #2?

**Comment by dhh1128** [comment]
A clarification: I see Kyle's comment in the stream about how to get sender anonymity. What I'm asking is whether "sender anonymity" is truly the same thing as "repudiable origin". Under some definitions, these might be synonyms; under others, maybe not. To me, "repudiable origin" means that nobody other than the sender and the receiver can know with certainty who the sender is--whereas sender anonymity could mean the sender is not known even to the receiver.

**Comment by dhh1128** [comment]
I have another requirement to posit:

3. Decryption cannot depend on DID resolution.

We can put a key ID in a `kid` field, and the value we put there can be a DID-based key reference. However, the decryption itself can't require a dereference, because we end up with a timing problem (the key value at the time of dereference isn't guaranteed to be the key value at the time of encryption). So we must be decrypting with something other than `kid`.

Is our solution doing that?

**Comment by Baha-sk** [comment]
```
{
    "protected": base64url({
        "typ": "prs.hyperledger.aries-anon-message",
        "alg": "ECDH-SS+XC20PKW",
        "enc":"XC20P"
    }),
    "recipients": [
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "apu": "base64url(random_bytes(64))",
                "nonce": "base64url(random_bytes(nonce_size))",
                "tag": "base64url(encrypted CEK tag)",
                "kid": "base64url(Recipient vk)",
                "spk": "EphemeralJWK(Sender vk, Recipient vk)",
            }
        },
        {
            "encrypted_key": "base64url(encrypted CEK)",
            "header": {
                "apu": "base64url(random_bytes(64))",
                "nonce": "base64url(random_bytes(nonce_size))",
                "tag": "base64url(encrypted CEK tag)",
                "kid": "base64url(Recipient vk)",
                "spk": "EphemeralJWK(Sender vk, Recipient vk)",
            }
        }
    ],
    "aad": "base64url(sha256(concat('.',sort([base58(recipients[0] vk), ..., base58(recipients[n] vk ])))))",
    "iv": "base64url(content encryption IV)",
    "ciphertext": "base64url(XC20P(DidExchange payload, base64Url(json($protected)+'.'+$aad), content encryption IV, CEK))"
    "tag": "base64url(AEAD Authentication Tag)"
}
```
Where `EphemeralJWK(Sender vk, Recipient vk)` outputs a full JWE format defined as follows:
```
base64Url(Headers).base64Url(encryptedKEK).base64Url(random_bytes(nonce_size)).base64Url(encryptedJWK cipher).base64Url(encryptedJWK tag)
```

where:
* `Headers` are:
```
	Headers: [
	        "typ":  "jose",
	        "cty": "jwk+json",
	        "alg": "ECDH-ES+XC20PKW",
	        "enc": "XC20P",
	        "epk": [
	        	"kty": "OKP",  // OKP not 0KP
	        	"crv": "X25519",
	        	"x": base64url($epk), // this is an ephemeral public key created to encrypt KEK
	        ],
	        "iv": base64url($knonce), // IV used to encrypt KEK
	        "tag": base64url($ktag), // tag of encyprtedKEK
	]
```

* `encryptedKEK` is: cipherText of `XC20P(CEK, '', $knonce encryption IV, $kek)`. 
	$kek is created as the result `$z=scalarmult($esk, recipient vk)` then `concatKDF($z, 'XC20PKW', 256)`where $esk  is an ephemeral secret key created with $epk as a key pair.

* finally `encryptedJWK tag`, `encryptedJWK cipher` are the encryption of the sender JWK defined as:
```
	 senderJWK:[
	        "kty": "OKP",  // OKP not 0KP
	        "crv": "X25519",
	        "x": base64url(Sender vk),
	 ]
```
with this call: `XC20P(senderJWK, base64Url(Headers), $nonce encryption IV, kek)`.

Example:
```
Sender SK   : W_WR77SQUhHikPHXn-CP5SynZ8ebMIh7SGe5GTEnbcw
Sender PK   : sapGsej_in9ISk06-2l9dimNHewl5f5-U9FKAnMcH2U
Recipient SK: Y92S7bhd8dJR-9X5yVrLUVlEH0vdNfd52No7aMftW7A
Recipient PK: a3iqc1ORk_y_HvfZtncBPxgWaVM3NA2cWLyf0OG9qQc
Payload     : SGVsbG8gV29ybGQh. // base64Url("Hello World!")
```
```
Headers: 
{"typ":"prs.hyperledger.aries-auth-message","alg":"ECDH-SS+XC20PKW","enc":"XC20P"}
```
```
Envelope:
{
    "protected": "eyJ0eXAiOiJwcnMuaHlwZXJsZWRnZXIuYXJpZXMtYXV0aC1tZXNzYWdlIiwiYWxnIjoiRUNESC1TUytYQzIwUEtXIiwiZW5jIjoiWEMyMFAifQ",
    "recipients": [
        {
            "encrypted_key": "rPy69f85-4oqdSF2pmOpxw6FXE9ID--XelyKndkGJ7k",
            "header": {
                "apu": "TlhMaHseS0urFH3qfwFhBluPBZekR6Ro_haJJvpxVs64bkXP5jH3-eGyirmgMkxr9FgAx6Zieau2EsseQs6qqg",
                "iv": "kWLiKhsER0FGgSahLjL4HAe5W8d98Ol_",
                "tag": "-d8xg42QPabb_oOkHxDcuw",
                "kid": "8EXLnYec1Lh2w92WSyo5KWaBwD3QaEKjzahFCp4QaTH8",
                "spk": "eyJ0eXAiOiJqb3NlIiwiY3R5IjoiandrK2pzb24iLCJhbGciOiJFQ0RILUVTK1hDMjBQS1ciLCJlbmMiOiJYQzIwUCIsImVwayI6eyJrdHkiOiJPS1AiLCJjcnYiOiJYMjU1MTkiLCJ4IjoiLVFOOFpLR2VrcTQyS09DcnVSRWdVZDk0YUFCd2l6alpHUzlmSWR1YXlVUSJ9LCJpdiI6IlgtaWx1SHFHOEhwbDFTMHNSQVR3MFlWYk1ZRDlVdHlIIiwidGFnIjoiRXpEbTl0cjdNOS1DOVRFNURseTc0dyJ9.6wzR556W_NtyPq7o55Giou0MJTGSfT1eYZhV2hv_GlE.3mB89GEslNpzkCo_5BrG36XlYE2xXeMY.QXwOr3AmH_Xinoe6-4fnhoUYrm3f9py8vAdTrua3BTutsYnZbr7OQIbSQqy4lO7TwsZ8KdJtfijFlbhdlrFUbPL3RdmqXcx3UGgw2FHg.vRe7YRKtYnAnDin9qwlPwA"
            }
        }
    ],
    "aad": "hgMzrJYQWNginfZSRF2dPEm01XkyEetXWJj_QDb5hmA",
    "iv": "g0Z7v-Fl02MgIZPdsEhDUp_Z4ZGvP0yk",
    "tag": "ioH7xaCdDSwzRG4Hhp1VHw",
    "ciphertext": "wOb0codq4GydjtqB"
}
```

PHP implementation:
https://github.com/gamringer/php-authcrypt
aries-framework-go Agent implementation:
https://github.com/hyperledger/aries-framework-go/pull/341

**Comment by dhh1128** [comment]
@Baha-sk This shows that `kid` contains a recipient's verkey. That's fine with me (it satisfies requirement 3 that I articulated [here](https://github.com/hyperledger/aries-rfcs/issues/133#issuecomment-535153437))--but I thought others were advocating that `kid` should contain a DID-based key reference instead. See https://github.com/hyperledger/aries-rfcs/issues/104#issuecomment-525396519

**Comment by Baha-sk** [comment]
@dhh1128 for now, the PHP and Go implementations are using recipient verkey as `kid` value.. the aim is this code is running within the wallet (as part of Pack()/Unpack()).

We probably can switch `kid` to use DID-based key references but I guess the code needs to be refactored to extract Pack/Unpack out of the wallet and call crypto code with key materials inside the wallet.. it will need further refactoring if we opt to go this route.

**Comment by dhh1128** [comment]
Right. I like that approach. It is what I was recommending. However, I got the impression that some in the community were advocating the opposite, so I was just noting that there may be pushback.

**Comment by Baha-sk** [comment]
I don't see any harm of sharing public keys in the envelope

**Comment by llorllale** [comment]
@Baha-sk 

> base64Url(Headers).**base64Url(encryptedKEK)**.base64Url(random_bytes(nonce_size)).base64Url(encryptedJWK cipher).base64Url(encryptedJWK tag)

Encrypted JWKs are JWEs with the entire JWK as the plaintext - see https://tools.ietf.org/html/rfc7517#appendix-C, not just the KEK.

**Comment by Baha-sk** [comment]
@llorllale the value of the recipient header's `spk` field is the JWE in compact format. The second field is the encryption key.

see the final section (9) of the same appendix-[C](https://tools.ietf.org/html/rfc7517#appendix-C.9)

**Comment by llorllale** [comment]
Right, my bad - I mistook the ordering of the segments.

**Comment by llorllale** [comment]
@dhh1128 

> We can put a key ID in a kid field, and the value we put there can be a DID-based key reference. > However, the decryption itself can't require a dereference, because we end up with a timing problem (the key value at the time of dereference isn't guaranteed to be the key value at the time of encryption). So we must be decrypting with something other than kid. This means kid is a hint or advisory field, useful for checking whether the decryption key has the DID context we're hoping for--but is never primary input to decryption.
> 
> Is our solution doing that?

We will add the DID-based key reference as a `kid` in the JWK that encodes the sender's verKey (the `spk` in Baha's snippets). Not to be confused with `recipient.header.kid`, which contains the recipient's inlined key.

**Comment by llorllale** [comment]
@dhh1128 can you elaborate on this point? I'm not sure what you're referring to:

> Solve the problem of runaway bloat caused by repeated re-encryption. (This is NOT an absolute dealbreaker for me, but is highly desirable. We proposed a few months back to change the encryption envelope to address this, but the proposal was deferred because we didn't think it was worth it to modify the envelope format solely to add this feature. Now that we're considering a change to the format for different reasons, I strongly prefer that we tackle this problem at the same time, and not defer it again.)

**Comment by dhh1128** [comment]
@llorlale: This sounds like an incomplete solution. Let me posit a scenario and then you can tell me if I'm understanding wrong.

Alice sends a message to Bob. When she sends the message on Monday, she uses one of her keys that is known as `did:example:alicedid#1`, with public key value `ABCXYZ`, and she encrypts FOR Bob's public key that is known as `did:example:bobdid#3`, with public key value `DEFPQR`. Thus, if I understand you right, the string `did:example:alicedid#1` gets put into `spk`, and the string `DEFPQR` gets put into `recipient.header.kid`. So far, do I have this right?

Now, Bob loses his phone in the couch cushions for 3 days, so he doesn't actually open Alice's message until Thursday. But on Wednesday, Alice rotated her key, such that `did:example:alicedid#1` now has public key value `GHIJKLM`.

When Bob opens the message, he can decrypt it immediately, because he sees `DEFPQR` (a key he knows about) in the `recipient.header.kid` field. Then, having decrypted the message, he decides to validate the sender. So he takes what he sees in `spk`, which is the string `did:example:alicedid#1`, resolves Alice's DID, and learns that it has the value `GHIJKLM`. When he checks to see if the key used to encrypt the message corresponds to that key, the check fails. It appears to Bob like Alice is lying.

What I think we need is the actual VALUE, not a REFERENCE, to the sender's public key at the time of message sending. We could *also* have a reference--but the value, not the reference is what needs to be used to do primary validation on the sender. Then, as a secondary step, it might be desirable to confirm that the sender's verkey did (at some timeframe in the past that roughly corresponded to send time) have the `id` property implied by the key reference hint.

**Comment by dhh1128** [comment]
@llorllale : The "runaway bloat" problem is pictured here: https://docs.google.com/presentation/d/1zE5YLzU2VUFdRy3lNpv4LOwMy15yJAkZMO_eitfNSOk/edit#slide=id.g4d0fe0a930_0_0. Take a minute to follow the link; I think a picture is worth a thousand words.

But here are some words to go with it: basically, every time you wrap content in an encryption transformation, you are re-encoding content. If a particular DIDComm channel has 5 hops, then that's 1 necessary encryption and 4 unnecessary ones. Not only does this waste CPU cycles, it also increases message sizes--maybe a lot.

What we'd like to do at each new wrapping is encrypt just the metadata from the previous message, but not re-encrypt the already encrypted payload.

**Comment by Baha-sk** [comment]
@dhh1128 I don't think we have a "runaway bloat" issue with our embedded jwk solution above. The encryption from our solution happens more like this:
```
Encryption 1 (main outer JWE):
	CEK --> enc --> Payload
Encryption 2 (recipient A CEK):
	KEKA --> enc --> CEK
Encryption 3 (sender's JWK as payload for Recipient A represented as an inner compact JWE):
	CEKA --> enc --> Sender JWK
Encryption 4 ( recipient A's sender CEK)
	SenderKEKA --> enc --> CEKA
```
Encryptions 2, 3 and 4 are repeated for as many recipients as we have on the list. 
We are not encrypting `Payload` or `CEK` over and over for each recipient.

**Comment by dhh1128** [comment]
@Baha-sk I'm not saying the bloat is affected by the number of recipients. I'm saying it's affected by the number of times re-encryption happens. What if I take a fully encrypted message (encrypted for, say, 3 recipients--or 1 or 20 or whatever) and encrypted it *again*, for 2 different recipients, and then encrypt it again for 1 yet different recipient?

I think you are assuming that plaintext is encrypted once, and that's it. I'm assuming that plaintext is encrypted for an *intermediate* recipient, who then decrypts and sees inside the decrypted envelope a payload that is encrypted for the next recipient down the line. This is an onion encryption pattern, and could be repeated many times. It happens in routing.

**Comment by Baha-sk** [comment]
@dhh1128 I believe this depends on how we handle routing.

If middle agents are `only routing` the message then these agents should probably not do Pack/Unpack as they will cause an `onion routing` effect as you described  above.

I think we need to update the RFC to mention how routing between agents should handle the payload it's receiving to avoid encrypting the payload over and over for each middle agent.

In my previous post I was referring to a single agent communication who is sending a payload to multiple recipients regardless of routing agents ( I did not know the list of recipients are basically `routing` agents + the final agent).

I do think each `routing` agent should simply pass through the envelope as-is so that the end recipient knows the route of the message. Is there an RFC for this?

But that doesn't change the scenario from my previous post, Encryption 1 happens only once, and the 3 other ones are for each recipient. 

Whether the sending agent knows the full route or only the end recipient and the next agent in the route, that's up for a separate discussion.

**Comment by dhh1128** [comment]
@Baha-sk : The routing behavior of DIDComm is already set, and it already includes onion routing, as spec'ed in several RFCs. See, for example, [RFC 0046](https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0046-mediators-and-relays) and [RFC 0094](https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0094-cross-domain-messaging). This is not going to change, and we can't ignore it in our encryption solution, or demand that it change for encryption convenience. There *will* be multiple layers of re-encryption. We can't only encrypt once, because then every intermediate router sees the final destination and can monitor where traffic is going--a privacy violation. This means that each hop in the journey of a message decrypts the metadata that tells it where to route for the next hop. Also, there is not just one final agent receiving a message. If Alice is the recipient, and she owns a laptop, an iPhone, and a tablet, then 3 agents will be the recipients at the end.

**Comment by Baha-sk** [comment]
@dhh1128 in this case you're right, each hop is responsible of encrypting (and wrapping) the envelope for the following hop in the route. Then we can't do without re-encrypting the envelop at each hop.


> Also, there is not just one final agent receiving a message. If Alice is the recipient, and she owns a laptop, an iPhone, and a tablet, then 3 agents will be the recipients at the end.

I believe the sender should/may not be aware of how many agents Alice has. A middle hop in the route is most probably aware of Alice's three agents via say an email account or a cloud service where Alice's devices are registered. This intermediate hop will encrypt and routes the message to all three agents using their corresponding verification keys.

**Comment by dhh1128** [comment]
>I believe the sender should/may not be aware of how many agents Alice has. A middle hop in the route is most probably aware of Alice's three agents via say an email account or a cloud service where Alice's devices are registered. This intermediate hop will encrypt and routes the message to all three agents using their corresponding verification keys.

If the final router before the ultimate destination is an agent of the final recipient, it *may* be acceptable for that router to see the message as plaintext and and re-encrypt it. However, generally this is not the case, and we certainly can't build a routing solution that depends on this behavior. The more common pattern, and the only safe default, is to assume that only the final recipients at the edge should see plaintext, and that the sender must therefore multiplex-encrypt for all the receiver's edge keys.

The sender does not know how many agents Alice has. However, the sender does have Alice's DID Doc, and does know how many *keys* Alice has that are not listed in endpoint definitions as intermediate routing keys. It is that set of keys that the sender must encrypt for in the recipients array.

**Comment by Moopli** [comment]
@Baha-sk 
> I believe the sender should/may not be aware of how many agents Alice has. A middle hop in the route is most probably aware of Alice's three agents via say an email account or a cloud service where Alice's devices are registered. This intermediate hop will encrypt and routes the message to all three agents using their corresponding verification keys.

I'm not sure how such an intermediate agent could reroute a message without either:
- the message being decryptable by all destination agents that it will reroute the message to
- the message being decryptable by the intermediate agent

For clarity: if A sends a message addressed to C, routed through B, with the expectation that B is a rerouting agent that will redirect the message to C1, C2, C3 etc, without A knowing the identities of C_1, C2, C3 etc, then either:
- C1, C2, C3 etc must be able to unpack a message addressed to C (ie, have the requisite keypair)
- B must be able to unpack a message (to cleartext, unpacking all layers) addressed to C so it can pack it for sending to C1, C2, C3 etc

For some uses (eg, a trusted cloud agent that you expect will do processing based on your data and also pass messages along) this is fine, but I don't think this is fine for every use case, though I defer to others to suggest the use cases.

**Comment by dhh1128** [comment]
@Moopli I agree with your summary of the issue. This is exactly the analysis that's described in detail in [RFC 0046](https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0046-mediators-and-relays). Mediators are NOT trusted with plaintext. Rather, they receive an envelope that contains an encrypted payload. The only thing the mediators can understand from looking at the envelope is that a routing request is being made, and the next destination in the route is X. Everything else is hidden.

**Comment by llorllale** [comment]
@dhh1128 you are correct with your https://github.com/hyperledger/aries-rfcs/issues/133#issuecomment-535691424 in that putting a reference to the Sender's key inside the `spk` and the moment of sending does not solve the timing problem.

The thing is I'm not trying to solve that problem - I'm attempting to solve the "reverse lookup" problem although in a sloppy way I admit. But now I went back to one of your comments on #104 and see that you had actually proposed to insert this claim as `sid` ("sender ID"). I'm fine with `sid`.

**Comment by Moopli** [comment]
@dhh1128 about onion routing bloat: from what I can tell the main cause of bloat is base64 encoding ciphertext for packing, then encrypting and b64encoding it again. Each encode is a 4/3 multiplier on ciphertext size. The rest of the bloat is from the (effectively constant) size of the additional JSON wrapping, headers, etc, and most of _that_ bloat remains with a header encryption chain, but that's definitely much more manageable (linear with onion layer depth rather than exponential). An ascii-safe encoding for binary data like the ciphertext is important for transmission, but an alternative solution would be:

1. Given an envelope P as payload to wrap inside another envelope E
2. Edit P by base64 decoding the ciphertext
3. encrypt P
4. base64 encode the resulting ciphertext
5. build E around that

Then when decrypting, the recipient will base64 decode and decrypt the ciphertext as usual, and then will see if it's an envelope to send on, in which case it base64 encodes just the ciphertext of this envelope.

You could split this between packing/unpacking and sending/receiving: when packing, don't b64 encode the ciphertext, and only b64 encode the ciphertext of the final envelope once all envelope layers have been packed. When receiving, b64 decode the ciphertext of the received envelope, and then unpack multiple layers if possible.

**Comment by dhh1128** [comment]
@lovesh , you may want to comment on @Moopli 's idea. (Lovesh had previously proposed a solution to this problem.)

**Comment by llorllale** [comment]
@dhh1128 @Baha-sk @kdenhartog @Moopli @lovesh 

I think it's a good idea to reduce the size of the `pack()` output. However, can we consider solving the "runaway bloat" problem related specifically for *routing* in a separate RFC? I think it would be cleaner if the re-encryption problem is addressed in the outer envelope (the `Forward` message). Yes, we would be re-encrypting the output of `pack()` but hopefully only **once**, and hopefully the metadata on this outer envelope is editable such that we can add hops as we please without having to touch the ciphertext again.

Just my 2 c.

**Comment by Moopli** [comment]
I should note that I'm not trying to propose this as an alternative - I want to point out that the repeated base64 encoding (really, any binary -> ascii-safe encoding, if repeated on data) is the source of the exponential bloat. In fact it occurs to me that my scheme runs into problems when decoding, since arbitrary binary data (the payload) could contain unescaped string-terminating characters and in general break json parsing (if your payload is an envelope, which you must parse as JSON to forward to the next recipient, and the ciphertext within that envelope is binary rather than for example base64).

@llorllale fair enough

**Comment by lovesh** [comment]
@Moopli As you pointed out in your last comment, having the ciphertext of the inner envelope as bytes is not safe (you could represent the bytes as integers but that will increase the size as well). Secondly, decoding the inner envelope from base64 each time before re-encoding the outer envelope is expensive.

**Comment by kdenhartog** [comment]
@lovesh can you referencing me to some documentation that shows that re-encrypting is not safe? I was not aware that this was a problem and would like to read up on it so I'm more aware of the problems it presents.

**Comment by Moopli** [comment]
[Here](https://github.com/hyperledger/aries-framework-go/issues/454)'s a concern we've been running into recently with aries-framework-go. The current envelope format uses signing keys everywhere, converting them to encryption keys (as you can do with ed25519/curve25519) solely for encryption. Note that this extends to packing signing keys in the envelope, with the recipient doing conversion as well. @Baha-sk's current Go implementation of this draft spec packs the encryption keys, and since it's interoperable with the PHP example I'd assume @gamringer is doing the same. 

There's a hole here:
- DIDexchange does not exchange encryption pubkeys in addition to verification keys
- This spec does not mandate that signing keys be used to create encryption keys, or that the keys included as part of the envelope be verification keys

Now, for RSA, for example, this should not be an issue, verification keys are encryption keys and could technically be used for both purposes, but for Curve25519, where we'd want to use Ed25519 for EdDSA, we do need to convert.

Personally I'd prefer that we generate signing and encryption keypairs as distinct keypairs, and develop a DIDexchange protocol that will exchange both, no matter whether the asymmetric crypto algorithm in use allows us to use the same keys or derive one from the other.

**Comment by dhh1128** [comment]
Agreed. I think @kdenhartog was planning to go this direction.

**Comment by llorllale** [comment]
@Moopli @dhh1128 @kdenhartog would it make sense to bake encryption keys into the `~service` decorator and DIDDoc `service` attribute by adding a note in the [didcomm conventions RFC](https://github.com/hyperledger/aries-rfcs/blob/master/features/0067-didcomm-diddoc-conventions/README.md)?

**Comment by kdenhartog** [comment]
> [Here](https://github.com/hyperledger/aries-framework-go/issues/454)'s a concern we've been running into recently with aries-framework-go. The current envelope format (I checked ACA-Py and Indy SDK) uses signing keys everywhere, converting them to encryption keys (as you can do with ed25519/curve25519) solely for encryption. Note that this extends to packing signing keys in the envelope, with the recipient doing conversion as well. @Baha-sk's current Go implementation of this draft spec packs the encryption keys, and since it's interoperable with the PHP example I'd assume @gamringer is doing the same.
> 
> There's a hole here:
> 
>     * DIDexchange does not exchange encryption pubkeys in addition to verification keys
> 
>     * This spec does not mandate that signing keys be used to create encryption keys, or that the keys included as part of the envelope be verification keys
> 
> 
> Now, for RSA, for example, this should not be an issue, verification keys are encryption keys and could technically be used for both purposes, but for Curve25519, where we'd want to use Ed25519 for EdDSA, we do need to convert.
> 
> Personally I'd prefer that we generate signing and encryption keypairs as distinct keypairs, and develop a DIDexchange protocol that will exchange both, no matter whether the asymmetric crypto algorithm in use allows us to use the same keys or derive one from the other.

What's being done in both the IndySDK implementations, the ACA-py implementation, and DIDComm-js is a conversion between an `ed25519` key to an `x25519` key. This is made possible by birational equivalence between Montgomery curves and twisted Edwards curves. 

See here for the math: https://en.wikipedia.org/wiki/Montgomery_curve#Equivalence_with_twisted_Edwards_curves

As for whether this is secure or not is another question. In the use of RSA keys, it's been proven that reuse of an RSA key for both signing and encryption is [insecure when using textbook RSA](https://crypto.stackexchange.com/a/12138). However, with elliptic curves, there's been no one to produce proof showing it's secure or insecure to reuse keys for DHKE and signing operations. With that said, I tend to agree that the use of independent keys is a better practice and would suggest that's the direction we go.

However, I will point out that doing this will introduce new complexity around how we're managing keys and would need some new did doc conventions to list these keys in the DID Doc.

> @Moopli @dhh1128 @kdenhartog would it make sense to bake encryption keys into the `~service` decorator and DIDDoc `service` attribute by adding a note in the [didcomm conventions RFC](https://github.com/hyperledger/aries-rfcs/blob/master/features/0067-didcomm-diddoc-conventions/README.md)?

This is another option to consider. It could be useful to do this, but I'm of the opinion that stuffing more of the cryptography into the semantic layer is an anti-pattern because it requires developers who want to focus on business logic and protocols to now how to learn about proper key rotation, key reuse, and the different tradeoffs between key types. Essentially, they have to learn enough crypto to be dangerous. Rather it's of my opinion that we contain the cryptography to the [secure channel protocol](https://www.youtube.com/watch?v=C-kWVb81tCc&feature=youtu.be&t=1758). Especially if we expect the use of multiple parties in the communication.

**Comment by Baha-sk** [comment]
@kdenhartog thanks for the detailed reply. I believe the envelope exchange is only used to encrypt/decrypt the packed payload, therefore it would make sense for the exchanged keys to be crypto keys and the payload should contain signing keys needed to verify its content.

So the message flow in an agent would look like this:
- The sending agent will:
1. Sign a payload and embed its public verification (signing) key in the payload (as a DID document)
2. Call Pack(payload above + encryption keys), no signing keys involved for Pack

- The receiving agent will:
3. Call Unpack(received envelope + their encryption key), no signing key involved here
4. Verify payload extracted from Unpack above using the embedded the corresponding recipient verification(signing) from the extracted payload (the receiving agent can fetch from their wallet a matching sender verification key).

This way, steps 3 and 4 above don't handle signing keys at the `Envelope` layer while the `payload` has all it needs for signing/verification.


The PHP example and Go example I posted earlier ([here](https://github.com/hyperledger/aries-rfcs/issues/133#issuecomment-535199768)) is using encryption keys only (no signing keys) in the envelope. This should be in concordance with your opinion.

By the way, I liked the video about the Noise protocol from the YouTube video you posted. We do implement some of the features in the framework in our example, but stick to JWE compliance (with ephemeral keys + nonce, the Noise framework states nonces are not needed with ephemeral keys, that's another topic)

**Comment by kdenhartog** [comment]
> 
> 
> @kdenhartog thanks for the detailed reply. I believe the envelope exchange is only used to encrypt/decrypt the packed payload, therefore it would make sense for the exchanged keys to be crypto keys and the payload should contain signing keys needed to verify its content.
> 
> So the message flow in an agent would look like this:
> 
>     * The sending agent will:
> 
> 
>     1. Sign a payload and embed its public verification (signing) key in the payload (as a DID document)
> 
>     2. Call Pack(payload above + encryption keys), no signing keys involved for Pack
> 
> 
>     * The receiving agent will:
> 
> 
>     1. Call Unpack(received envelope + their encryption key), no signing key involved here
> 
>     2. Verify payload extracted from Unpack above using the embedded corresponding recipient verification(signing) from the extracted payload (the receiving agent can fetch from their wallet a matching sender verification key).
> 
> 
> This way, steps 3 and 4 above don't handle signing keys at the `Envelope` layer while the `payload` has all it needs for signing/verification.
> 
> The PHP example and Go example I posted earlier ([here](https://github.com/hyperledger/aries-rfcs/issues/133#issuecomment-535199768)) is using encryption keys only (no signing keys) in the envelope. This should be in concordance with your opinion.
> 
> By the way, I liked the video about the Noise protocol from the YouTube video you posted. We do implement some of the features in the framework in our example, but stick to JWE compliance (with ephemeral keys + nonce, the Noise framework states nonces are not needed with ephemeral keys, that's another topic)

This is a good idea. In the JWS, we can provide a reference to a different key than the one referenced in the outer JWE layer. This raises the question though, what happens when the key used to sign the JWS is in a different DID Doc than the one in the JWE? There's some weird edge cases we should be accounting for here, but in total - I agree this is a good solution.

**Comment by dhh1128** [comment]
I want to re-emphasize something about key references that I've said before (possibly on different discussion threads): *crypto can't depend on DID resolution*. We cannot be looking up key references in DID documents to do either of the following operations:

* decrypt
* validate a signature

The value of the `kid` field in JW* structures is not valid input to either of these operations. We *could* use a key reference to confirm that the key we're seeing has proper standing with respect to the party that's purporting to communicate with us--but that is not the same thing as the cryptographic operation of validating a signature.

Are we aligned on this, or did I just raise eyebrows?

**Comment by Baha-sk** [comment]
@dhh1128 that's what I'm trying to get at, JWE must not have any verification key stored in its headers. The keys stored in the envelope should be for encryption/decryption of the JWE only. This is why we should push the RFC to use the new JWE format I was referring to in my previous post.

**Comment by dhh1128** [comment]
@Baha-sk : I am content with the idea that we should split verification keys and encryption keys apart; the conflation of these was an accident of us using Ed25519. And I get that you're advocating the removal of the verkeys from the JWE. That's all fine. But what I don't know if we're aligned on is this: I want an encryption key *value*, NOT a *reference* to an encryption key named in a DID doc--to be named in the JWE, and I want a verkey value--NOT a reference to a verkey in a DID doc--as the field that a JWS uses to verify.

**Comment by Baha-sk** [comment]
@dhh1128 the encryption keys in the JWE example are real public key values. They are not reference IDs nor are they stored in the DID doc.

**Comment by troyronda** [comment]
Saw this update on the DIF issue tracker:
https://github.com/decentralized-identity/DIDComm-js/issues/8#issuecomment-545125297

(referring to https://github.com/digitalbazaar/minimal-cipher)

**Comment by troyronda** [comment]
The above effort allows:

- content encryption algorithms:
 - ChaCha20-Poly1305 (C20P)
 - 256-bit AES-GCM (to allow for a FIPS compatibility mode)
 - future: XChaCha20-Poly1305 (XC20P)

- key encryption algorithms:
  - ECDH-ES+A256KW

Also note the kid format:
> ${did}#${keyAgreementKey.fingerprint()}

**Comment by dhh1128** [comment]
The kid format is fatally flawed unless it includes a versionstamp or (less optimally) a timestamp. How can we get it upgraded?

**Comment by troyronda** [comment]
Yes. It would be nice to have better alignment on these topics.

I was also going to mention that it would be nice to have overlapping algorithm support in the envelope. Note that they have `ECDH-ES+A256KW` and additionally allow for 256-bit `AES-GCM`. The rationale for the latter also seems like a good idea (allowing for FIPS).

**Comment by tplooker** [comment]
How is the `kid` fatally flawed? In an encrypted envelope the kid is just suppose to act as an alias to a private key for the recipient, DID resolution isn't even required. The kid can literally be anything, as long as it is meaningful to the recipient, it really comes down to what the recipient indexes the keys in their kms against.

**Comment by tplooker** [comment]
I think we need to be clearer around the intention of the kid in different circumstances. 

When we are talking about encryption/decryption, the kid is really acting as an alias to a private key the recipient should have, whereas when we are verifying a signature, the kid should allow the audience of the signature to obtain a public key to perform verification.

**Comment by dhh1128** [comment]
I agree that we need to be clearer about intention. However, that doesn't affect my assertion that the `kid` format of `${did}#${keyAgreementKey.fingerprint()}` is fatally flawed.

Understand: I'm not saying that the *idea* of `kid` is fatally flawed. I'm staying that the specific `kid` *format* is.

It's flawed because without a versionstamp or timestamp, it can't be dereferenced -- either for encryption or for signature validation.

Analogy: you want an email address, and I give you "@example.com" as the answer. I claim that answer is fatally flawed, because it doesn't include the mailbox identifier.

A `kid` without a versionstamp or timestamp *might* be dereferencable, sometimes, if no key rotations have ever happened. And people who are building trivial, non-production systems get away with using such a `kid` because they ignore this. But to *really* use a `kid` to solve production problems, you can't ignore it. If you do, your `kid` is fatally flawed.

**Comment by tplooker** [comment]
I understand your reasoning, but _this_ key format proposed I dont think is fundamentally flawed in the way you are describing for encryption and decryption, which is the context it is being applied in, because the fragment portion is a fingerprint. Which implies an immutable and unique reference to a single key. 

Say you rotate a key but some sender continues to send you encrypted messages to your old key, provided you still have the old key indexed in your kms by this kid format, you should still be able to decrypt the message. And safely decrypt messages from new keys linked to the same did.

I do understand that because identifiers for keys in did docs are not enforced as immutable by default (e.g not every key featured in a did doc has an id corresponding to the keys unique fingerprint) this is not a universal rule for all did methods. However, this as a kid format if it were universally supported I believe would work without major issue for encryption/decryption.

Essentially when it comes to a kid for a signature, the easiest option to preserve verification is to make the kid the public key, no dereferencing required. However when it is an encrypted message the kid must be some meaningful alias to the recipient, there really is no need to have a versionstamp or timestamp, either the alias means something to me as a recipient or it doesn't.

**Comment by dhh1128** [comment]
> I dont think is fundamentally flawed in the way you are describing for encryption and decryption, which is the context it is being applied in, because the fragment portion is a fingerprint. Which implies an immutable and unique reference to a single key.

The problem is that the key *value* is immutable, but the key's meaning with respect to the DID doc -- the context in which that key is related to the DID -- is not. That key value might have been authorized to encrypt on behalf of the DID at point A, and not authorized to encrypt at point B, and you wouldn't know that. So you could do naked decryption with just the key value, but you can't do the semantic validation that should follow naked decryption, without the additional context.

>However when it is an encrypted message the kid must be some meaningful alias to the recipient, there really is no need to have a versionstamp or timestamp, either the alias means something to me as a recipient or it doesn't.

I disagree. You are assuming that all decryption happens at the time of transmission, which is also the time of reception. If you separate those timeframes, the "either it means something to me or it doesn't" unravels. It might not mean something to me at the time I receive it, but it might be important that it means something to me later. Or vice versa.

**Comment by selfissued** [comment]
A `kid` value is used to match a particular key.  No less and no more.  They're assigned by the party that creates the key and they're simply strings that are intended to uniquely identify that key.  They work great for that purpose and are in widespread use.

**Comment by dhh1128** [comment]
@selfissued : Right. 100% agree. I think `kid`s are great.

My concern is that the intersection between `kid`s and DIDs is suffering because of a design decision in the DID spec (which I vehemently disagreed with but failed to change) that *the relationship between a key identifier and a key value inside a DID doc is mutable.* It is illegal to assume that did:foo:123#identifier always refers to the same thing.

Now, fans of JW* in DID circles have suggested eliminating `id` as a required property on a key, claiming that `kid` is an equivalent. I don't think they realize the mess they're signing up for. An `id` has meaning in the larger context of the rest of the doc, and one of those meanings, according to the DID spec, is that you can use it to construct DID URLs with particular semantics. Among those semantics is the idea of mutability--that the data behind the identifier can be rotated. So if `kid` is an equivalent of `id`, then it follows that a `kid` that's a DID URL composed of a DID doc plus an identifier is not allowed to be assumed to be stable with respect to time. This is in direct conflict with the intention of `kid`, which is the simple and stable mapping you articulated.

Possible solution 1: we stop thinking that `kid` and `id` mean the same thing. If someone wants a key to have a mutable relationship between its identifier and its value, they use `id`; if they want immutability, they use `kid`. (Personally, I would never use 'id'...) We don't confuse them. A DID URL in the format did:foo:123#xyz is an `id`-based URL, NOT a `kid`-based URL. If we want a `kid`-based URL, we use something different, like maybe did:foo:123;kid=xyz.

Possible solution 2: we like `kid` = `id`, and we accept that this means DID URLs that reference keys by `kid` must have a versionstamp or timestamp to disambiguate. We stop kidding ourselves that did:foo:123#identifier is enough to accomplish what `kid` does.

**Comment by tplooker** [comment]
I would like to mitigate the amount of complexity we try to pack into the `kid` field, I think the field should be just a simple alias to a key that either means something to me as a recipient or it doesn't. 

The case you describe @dhh1128 should be offloaded for the recipient software to deal with, e.g if recipient software rotates a key and the same kid is used in a did doc, what does the recipient software do with the old key? Delete it? If thats the case you can't decrypt anyway no matter if you have a timestamp or not. If the software keeps the key, then what is its index in the kms? How come it cannot preserve an association to the `kid` it formally represented? Then if you receive a message that is encrypted to a kid where there have been multiple keys represented by that `kid` overtime, you try them until you can successfully decrypt.

**Comment by tplooker** [comment]
Alternatively if you don't want to traverse previous keys for a `kid`, we could add an optional JOSE header field that describes when the message was encrypted, that would then give the recipient further clue as to which key they need to use to decrypt.

**Comment by tplooker** [comment]
I do however disagree with trying to pack more meaning into the `kid` field, like a timestamp/versionstamp

**Comment by troyronda** [comment]
> Alternatively if you don't want to traverse all previous keys associated to a `kid`, we could add an optional JOSE header field that describes when the message was encrypted, that would then give the recipient further clue as to which key they need to use to decrypt.

I assume you are thinking the JWT `iat` field?

**Comment by andrewwhitehead** [comment]
The DID spec provides a format for identifying a DID Doc at a specific time using either the `version-id` or `version-time` parameters, doesn't that satisfy the need to refer to a specific key? The `hl` property (hash) can also be used to ensure the document integrity.

`"kid": "did:method:ident;version-id=GJdXXRq7Kxu7uKQCb77ik3;hl= zQmWvQxTqbG2Z9HPJgG57jjwR154cKhbtJenbyYTWkjgF3e#key-ident"`

**Comment by tplooker** [comment]
It does, but doing this adds a lot of complex in parsing this field. `kid` is simply mean to be a string alias or a hint to a key, by making the requirement in DIDComm that this is a full URL we now need a url parser to extract this information and will need to evolve this syntax as DID resolution continues to involve.

If there is a problem with kid reuse for multiple keys over time I think the simple choice is for the audience or consumer of the message to iterate through the keys associated to one kid or for there to be a seperate field in the kid header that hints at timing information and therefore allows them to limit the possible keys the `kid` is referring too.

**Comment by kdenhartog** [comment]
IIRC, I heard there was discussion about this during the connectathon. Would anyone be able to provide details about that discussion here for historical context?

**Comment by dhh1128** [comment]
My memory of the conversation was that we got to a stalemate. @tplooker , please confirm. The comment thread here boils down to 2 basic positions, which also summarize the connectathon discussion:

1. (Tobias) It's onerous and ugly to make a DID key reference in `kid` include a timestamp or version, so let's not do that.
2. (me) If we use a DID key reference as the value of `kid`, it *must* include a timestamp or version. (The assertion that we could have an ambiguous `kid` that can be disambiguated by iterating over candidates until we find a match strikes me as worse than a code smell--it feels like a security vulnerability.)

FWIW, the [ticket about this issue in the DID spec repo](https://github.com/w3c/did-core/issues/131) includes a comment from one of the JWE spec authors (Mike Jones) that suggests the intent of `kid` is to be unambiguous--else its use in caching is impossible.

**Comment by tplooker** [comment]
@dhh1128, yes your summary is correct.

**Comment by TelegramSam** [comment]
Continue issue here: https://github.com/decentralized-identity/didcomm-messaging/issues/22
